api/session.ts

```diff
@@ -0,0 +1,442 @@
+		const rows = await users.database.get(
+			"SELECT COUNT(*) AS `registrationcount` FROM `ntbb_users` WHERE `ip` = ? AND `registertime` > ?",
+			[ip, timestamp]
+		);
+		if (!rows) return 0;
+		return rows['registrationcount'];
+	}
+	async addUser(username: string, password: string) {
+		const hash = await bcrypt.hash(password, Config.passwordSalt);
+		const userid = toID(username);
+		const exists = await users.get(['userid'], toID(username));
+		if (exists) return null;
+		const ip = this.dispatcher.getIp();
+
+		const result = await users.insert({
+			userid, username, passwordhash: hash, email: null, registertime: time(), ip,
+		});
+
+		if (!result.affectedRows) {
+			throw new Error(`User could not be created. (${userid}, ${ip}`);
+		}
+		return this.login(username, password);
+	}
+	async login(name: string, pass: string) {
+		const curTime = time();
+		this.logout();
+		const userid = toID(name);
+		const info = await users.get('*', userid);
+		if (!info) {
+			// unregistered. just do the thing
+			return this.dispatcher.user;
+		}
+		// previously, there was a case for banstate here in the php.
+		// this is not necessary, as getAssertion handles that. Proceed to verification.
+		const verified = await this.passwordVerify(name, pass);
+		if (!verified) {
+			throw new ActionError(`Wrong password.`);
+		}
+		const timeout = (curTime + SID_DURATION);
+		const ip = this.dispatcher.getIp();
+		const sidhash = await this.getSid();
+		const res = await sessions.insert({userid, sid: sidhash, time: time(), timeout, ip});
+		this.sidhash = sidhash;
+		this.session = res.insertId || 0;
+		this.dispatcher.user.login(name);
+		this.updateCookie();
+		return this.dispatcher.user;
+	}
+	async logout() {
+		if (!this.session) return false;
+		await sessions.delete(this.session);
+		this.sidhash = '';
+		this.session = 0;
+		this.deleteCookie();
+		this.dispatcher.user.logout();
+	}
+	updateCookie() {
+		const name = this.getName();
+		if (toID(name) === 'guest') return;
+		if (!this.sidhash) {
+			return this.deleteCookie();
+		}
+		const rawsid = encodeURIComponent([name, this.session, this.sidhash].join(','));
+		this.dispatcher.setHeader(
+			'Set-Cookie',
+			`sid=${rawsid}; Max-Age=31363200; Domain=${Config.routes.root}; Path=/; Secure; SameSite=None`
+		);
+	}
+	async getAssertion(
+		userid: string,
+		challengekeyid: number = -1,
+		user: User | null,
+		challenge: string = '',
+		challengeprefix: string = ''
+	) {
+		if (userid.startsWith('guest')) {
+			return ';;Your username cannot start with \'guest\'.';
+		} else if (userid.length > 18) {
+			return ';;Your username must be less than 19 characters long.';
+		} else if (!Session.isUseridAllowed(userid)) {
+			return ';;Your username contains disallowed text.';
+		}
+		let data = '';
+		const ip = this.dispatcher.getIp();
+		let forceUsertype: string | false = false;
+		if (!user) user = this.dispatcher.user;
+		if (ip.includes(Config.autolockip) || Config.autolockip?.includes(ip)) {
```

Comment:
Isn't `Config.autolockip` sometimes an array? If so, this code will autolock all users if that array is empty.
![image](https://user-images.githubusercontent.com/56906084/128565142-bdec603b-f966-4ff0-b545-2f5958b2d1d9.png)